### DISCUSSION TITLE
Splitting methods in separate files

### DISCUSSION THREAD

**Discussion by eNeRGy164** [fluentassertions/fluentassertions#1551] [Ideas]
Just a suggestion.

As I am currently working in `TypeAssertions.cs` (1500+ lines) and `TypeAssertionSpecs.cs` (5500+ lines) these files are huge.

Would it be an idea to split these files in separate files for maintainability?
Making the static class `partial` should not change the public signature.

I do this for my projects that contain large amounts of extension methods, f.e. [PlantUml.Builder/StringBuilderExtensions](https://github.com/eNeRGy164/PlantUml.Builder/tree/main/src/PlantUml.Builder/StringBuilderExtensions), [PlantUml.Builder.Tests/StringBuilderExtensions](https://github.com/eNeRGy164/PlantUml.Builder/tree/main/tests/PlantUml.Builder.Tests/StringBuilderExtensions).

Depending on the file structure I have a dedicated folder per class, or I name my files like `StringBuilderExtensions.SkinParameter.cs` containing all the `.SkinParameter(...)` methods and overloads.

Some of the benefits:
* Smaller files
* With changes, the filename gives context (a change in `TypeAssertions.cs` does not reflect which of the 55! methods is modified, a change in `TypeAssertions.HaveProperty.cs` does give it away)
* You can have methods open side-by-side in the IDE
* Comparisons/changes are easier shown (f.e. in GitHub)
* No regions needed anymore

**Comment by jnyrup** [comment]
I'm very positive.
It would also make history search easier with the fewer changes per file.

For our unit tests I've also thought about the approach described in http://haacked.com/archive/2012/01/02/structuring-unit-tests.aspx/ where each method under test has it's own subclass.
With the functional naming style we practice I sometimes have to get "creative" with the method naming to avoid conflicts.
E.g. `When_passing_in_a_null_sequence_it_should_throw` could then be re-used for many tests asserting on parameter validation.

**Comment by dennisdoomen** [comment]
I'm not very fond of partial files for the production code, especially because I always use Rider's Structure Window to see where a new method should go, or use it for reordering. However, I do acknowledge the problem with our test files. There it may make sense to group them in a better way or use partial classes.

**Comment by jnyrup** [comment]
Okay, let's try it out with test classes for now.

Here are some suggestions for the structure

1)
```
Types\
  TypeAssertionsSpecs.cs
  TypeAssertionsSpecs.Be.cs (both Be and NotBe in here)
```


2) Affirmative and negative assertions in separate files
```
Types\
  TypeAssertionsSpecs.cs
  TypeAssertionsSpecs.Be.cs
  TypeAssertionsSpecs.NotBe.cs
```

3) Subfolder per class.
E.g. GenericCollectionAssertions will have 44 different partial classes
```
Types\
  TypeAssertions\
    TypeAssertionsSpecs.cs
    TypeAssertionsSpecs.Be.cs
    TypeAssertionsSpecs.NotBe.cs
```

I would vote against simply naming them `Be.cs` as that would give us 18 classes with that name, which would harm the "goto file" experience in any IDE.

**Reply by dennisdoomen** [reply]
Since I don't like too many files either, I vote for 1.

**Reply by eNeRGy164** [reply]
Well, just split a commit with a hundreds of changes... I really opt for # 2 😀

**Reply by dennisdoomen** [reply]
Haha. Well, that's why I think option 1 is good enough. It keeps the positive and negative tests together, but still splits them by API.

**Reply by eNeRGy164** [reply]
I'd love to pick it up, but maybe my current PR is already getting a bit out of scope i think 😄

**Reply by eNeRGy164** [reply]
I would also expect the generic versions to be combined with the non generics as well for now. Although '`1' is allowed as file name.

**Reply by dennisdoomen** [reply]
> I'd love to pick it up, but maybe my current PR is already getting a bit out of scope i think

Nothing stops you from creating more PRs ;-)

**Reply by eNeRGy164** [reply]
Well, not if it's splitting up the file that is in my other PR. I will happily wait till that one is merged.

**Comment by eNeRGy164** [comment]
Funny, I see `DataEquivalencySpecs` is already doing this

**Reply by dennisdoomen** [reply]
Yeah, that was an external contributor. Since we didn't align on our "convention" yet, we accepted it.

**Comment by dennisdoomen** [comment]
So did we settle on option 1 @jnyrup @eNeRGy164 ?

**Reply by eNeRGy164** [reply]
Yes, option 1

See:
- https://github.com/fluentassertions/fluentassertions/tree/master/Tests/FluentAssertions.Specs/Types
- https://github.com/fluentassertions/fluentassertions/tree/master/Tests/FluentAssertions.Specs/Collections

**Reply by jnyrup** [reply]
yes

**Comment by dennisdoomen** [comment]
Then we also need to think of a way to improve the equivalency specs where all of them use the same method, `BeEquivalentTo`. We currently use regions, but that doesn't help in the test explorer. I've tried using a nested class, e.g. `Mapping` and that seems to work well in Rider.

![image](https://user-images.githubusercontent.com/572734/150633382-6cfc678f-91ae-47cd-b5b4-77d31a84fa2d.png)

**Reply by jnyrup** [reply]
Using nested classes also helps reducing the headache of duplicate method names.
